dasharo-security/secure-boot.robot - UEFI Shell less SBO tests

matmacieje opened the pull request:
SBO003.001 & SBO004.001 - WiP.
lib/secure-boot-lib.robot keywords for setup menu Secure Boot testing platform-configs/include/default.robot - output strings expected by SBO tests in menu-based execution

DonVasyl committed:
dasharo-security/secure-boot.robot - UEFI Shell-less SBO tests

Test cases SBO003.001, SBO004.001, SBO005.001, SBO007.001, SBO008.001
working without UEFI Shell dependency (100% setup menu based execution).
USB stick image labeled BAD_INFLUE is now required, containing new
hello_dasharo.efi test program, to enable serial port message capture.
Works with Telnet + physical USB stick and with PiKVM + image mount.

Signed-off-by: Mateusz Maciejewski <mateusz.maciejewski@3mdeb.com>

miczyg1 commented on dasharo-security/secure-boot.robot:
```diff
@@ -127,46 +127,41 @@ SBO002.002 UEFI Secure Boot (Windows)
-    Enter Enroll DB Signature Using File In DB Options    ${advanced_menu}
-    Enter Volume In File Explorer    GOOD_KEYS
-    Select File In File Explorer    DB.cer
+    # {advanced_menu}=    Enter Advanced Secure Boot Keys Management And Return Construction    ${sb_menu}
+
+    # TBD:
+
+    # Enter Enroll DB Signature Using File In DB Options    ${advanced_menu}
+    # Enter Volume In File Explorer    GOOD_KEYS
+    # Select File In File Explorer    DB.cer
     # Save Changes And Reset
     # Changes to Secure Boot menu take action immediately, so we can just reset
```
I am not sure if I should review further after seeing this commented code

matmacieje replied:
Done in https://github.com/Dasharo/open-source-firmware-validation/pull/725/commits/e9fbf1adbe7b5cf30c54aaae6c354f4852de20b1

matmacieje commented:
Squashed & rebased to develop.

miczyg1 commented on dasharo-security/secure-boot.robot:
```diff
@@ -263,3 +256,14 @@ Set Secure Boot State To Disabled
     # Changes to Secure Boot menu take action immediately, so we can just reset
     Tianocore Reset System
+
+Boot EFI Executable From The Drive
```
Why is it not a part of lib/secure-boot-lib.robot?

matmacieje replied:
Removed in https://github.com/Dasharo/open-source-firmware-validation/pull/725/commits/e9fbf1adbe7b5cf30c54aaae6c354f4852de20b1

miczyg1 commented on lib/secure-boot-lib.robot:
```diff
@@ -182,10 +202,11 @@ Enter Volume In File Explorer
     [Arguments]    ${target_volume}
     # 1. Read out the whole File Explorer menu
-    ${volumes}=    Get Submenu Construction    opt_only=${TRUE}
+    Sleep    1s
```
Why is the sleep needed? `Get Submenu Construction` should read everything util a checkpoint is found

matmacieje replied:
Removed in https://github.com/Dasharo/open-source-firmware-validation/pull/725/commits/e9fbf1adbe7b5cf30c54aaae6c354f4852de20b1

miczyg1 commented on dasharo-security/secure-boot.robot:
```diff
@@ -25,8 +25,6 @@ Suite Setup         Run Keywords
 ...                     Skip If    not ${SECURE_BOOT_SUPPORT}    Secure Boot is not supported
 ...                     AND
-...                     Mount USB Disk Image    ${TEST_DATA_DIR}/secure-boot/sb_test_data.img
-...                     AND
```
I believe we should not remove it. It used to mount an USB image via PiKVM with the same content as BAD_INFLUE to test Secure Boot. it should be possible to maintain the compatibility with PiKVM.

matmacieje replied:
Fixed in https://github.com/Dasharo/open-source-firmware-validation/pull/725/commits/e9fbf1adbe7b5cf30c54aaae6c354f4852de20b1

miczyg1 commented on dasharo-security/secure-boot.robot:
```diff
@@ -25,7 +25,7 @@ Suite Setup         Run Keywords
 ...                     Skip If    not ${SECURE_BOOT_SUPPORT}    Secure Boot is not supported
 ...                     AND
-...                     Mount USB Disk Image    ${TEST_DATA_DIR}/secure-boot/sb_test_data.img
+...                     Run Keyword If    "${DUT_CONNECTION_METHOD}" != "Telnet"    Mount USB Disk Image    ${TEST_DATA_DIR}/secure-boot/sb_test_data.img    file    TRUE
```
I am not sure this is correct, because even if we use `'${DUT_CONNECTION_METHOD}' == 'pikvm'` we still connect with Telnet for serial, e.g. on MSI to read the output. I am not entirely sure how this works and how the switching of DUT connection works. Is DUT connection switching done in Suite Setup? Probably not?

Generally my point is that mounting the USB disk image should be only done when PiKVM is present. Maybe this will work:

```suggestion
...                     Run Keyword If    '${INITIAL_DUT_CONNECTION_METHOD}' == 'pikvm' or '${DUT_CONNECTION_METHOD}' == 'pikvm'   Mount USB Disk Image    ${TEST_DATA_DIR}/secure-boot/sb_test_data.img    file    TRUE
```

matmacieje replied:
I can do this, but Mount USB Disk Image keyword serves also image mounts on QEMU platforms and I think it may not work with 'pikvm' value.

matmacieje replied:
Fixed in https://github.com/Dasharo/open-source-firmware-validation/pull/725/commits/eae118863a9ce586cac58720aa5d5fe8606e4cdf and tested on PiKVM & Telnet-connected DUTs

SebastianCzapla reviewed (changes requested):
It might be worth including names and keyword tags according to new naming conventions, to include environmental IDs in the test ID, and at the same time dropping "(Windows)" / "(Ubuntu)" / "(Firmware)". Details can be found in this repository at docs/test-naming-conventions.md.

SebastianCzapla commented on dasharo-security/secure-boot.robot:
```diff
@@ -251,8 +258,8 @@ SBO008.001 Attempt to enroll the key in the incorrect format (firmware)
+    Select File In File Explorer    cert_fake.der
     Read From Terminal Until    ERROR: Unsupported file type!
 
 
```
Could you also update the "Set Secure Boot State To Disabled" KWD to include 
```
[Tags]   robot:private
```
according to docs/contributing.md, KWDs local to a test suite, should be tagged in such way

matmacieje replied:
Fixed in https://github.com/Dasharo/open-source-firmware-validation/pull/725/commits/7e487b76c4d5bf47e1be1709ac59fc85cd6a1d8a

SebastianCzapla commented:
[secure-boot.robot.zip](https://github.com/user-attachments/files/19487757/secure-boot.robot.zip)
Here is a log from testing on VP3230 platform.

macpijan commented:
> It might be worth including names and keyword tags according to new naming conventions, to include environmental IDs in the test ID, and at the same time dropping "(Windows)" / "(Ubuntu)" / "(Firmware)". 

Let's not mix things up here.

We will do this when preparing tests for Fedora (and testing things).

DonVasyl committed:
secure-boot.robot Set Secure Boot State To Disabled tagged as private

Tag robot:private added to the local keyword.

Signed-off-by: Mateusz Maciejewski <mateusz.maciejewski@3mdeb.com>

DonVasyl committed:
secure-boot.robot: DUT-via-Telnet condition removed from image mount

lib/usb-hid-msc-lib.robot: added PiKVM & QEMU check in Mount USB
Disk Image keyword.

Signed-off-by: Mateusz Maciejewski <mateusz.maciejewski@3mdeb.com>
